src/dawn/Optimizer/PassMultiStageSplitter.cpp

```diff
@@ -27,101 +27,169 @@
 
 namespace dawn {
 
-PassMultiStageSplitter::PassMultiStageSplitter() : Pass("PassMultiStageSplitter") {}
-
-bool PassMultiStageSplitter::run(
-    const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
-  OptimizerContext* context = stencilInstantiation->getOptimizerContext();
-
-  int numSplit = 0;
-  std::deque<MultiStage::SplitIndex> splitterIndices;
-  DependencyGraphAccesses graph(stencilInstantiation.get());
-
-  // Iterate over all stages in all multistages of all stencils
-  for(auto& stencil : stencilInstantiation->getStencils()) {
-
-    int multiStageIndex = 0;
-    for(auto multiStageIt = stencil->getMultiStages().begin();
-        multiStageIt != stencil->getMultiStages().end(); ++multiStageIndex) {
-      MultiStage& multiStage = (**multiStageIt);
-
-      splitterIndices.clear();
-      graph.clear();
-
-      // Loop order specified by the user in the vertical region (e.g {k_start, k_end} is forward)
-      auto userSpecifiedLoopOrder = multiStage.getLoopOrder();
+PassMultiStageSplitter::PassMultiStageSplitter(MulitStageSplittingStrategy strategy)
+    : Pass("PassMultiStageSplitter"), strategy_(strategy) {
+  isDebug_ = true;
+}
+namespace {
+std::function<void(std::list<std::shared_ptr<MultiStage>>::iterator&, OptimizerContext*, int&,
+                   DependencyGraphAccesses&, std::shared_ptr<Stencil>, int&, std::string&)>
+setOptimizedLoopContent(const std::shared_ptr<StencilInstantiation>& stencilInstantiation) {
+  return [&](std::list<std::shared_ptr<MultiStage>>::iterator& multiStageIt,
+             OptimizerContext* context, int& numSplit, DependencyGraphAccesses& graph,
```

Comment:
context is not really needed, we can simply pass the boolean flags used, 
doDump...